sdl-core/src/main/scala/io/smartdatalake/workflow/action/spark/transformer/SparkFlattenDfTransformer.scala

```diff
@@ -0,0 +1,95 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.smartdatalake.workflow.action.spark.transformer
+
+import com.typesafe.config.Config
+import io.smartdatalake.config.SdlConfigObject.{ActionId, DataObjectId}
+import io.smartdatalake.config.{FromConfigFactory, InstanceRegistry}
+import io.smartdatalake.util.hdfs.PartitionValues
+import io.smartdatalake.workflow.ActionPipelineContext
+import io.smartdatalake.workflow.action.generic.transformer.{GenericDfTransformer, SparkDfTransformer}
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.functions.{col, explode_outer, get}
+import org.apache.spark.sql.types.{ArrayType, DataType, StructType}
+
+import scala.annotation.tailrec
+
+/**
+ * Returns a flattened Dataframe from another DataFrame
+ * with a nested schema. Schemas in the form:
+ *
+ * |---parent
+ *
+ * |------|child1
+ *
+ * |------|child2
+ *
+ * are transformed to the following flat columns:
+ *
+ * |---parent_child1
+ *
+ * |---parent_child2
+ * @param name         name of the transformer
+ * @param description  Optional description of the transformer
+ * @param enableExplode Enables exploding Arrays in the Dataframe, therefore potentially increasing the number of rows. Default is set to "false"
+ */
+case class SparkFlattenDfTransformer(override val name: String = "sparkFlattenDataFrame",
+                                     override val description: Option[String] = None,
+                                     enableExplode: Boolean = false) extends SparkDfTransformer {
+
+  private def getComplexFields(df: DataFrame): List[(String, DataType)] = {
+    enableExplode match {
+      case true => {
+        (for (field <- df.schema.fields
+              if (field.dataType.isInstanceOf[StructType] || field.dataType.isInstanceOf[ArrayType]))
+        yield (field.name, field.dataType)).toList
+      }
+      case false => {
+        (for (field <- df.schema.fields
+              if (field.dataType.isInstanceOf[StructType]))
+        yield (field.name, field.dataType)).toList
+      }
+    }
+  }
+
+  @tailrec
+  private def flattenDf(df: DataFrame): DataFrame = {
+    val complexFields = getComplexFields(df)
+    if (complexFields.isEmpty) return df;
+    val (colName, colType) = complexFields.head
+    val newDf = colType match {
+      case StructType(fields) => {
+        val inner = for (n <- fields) yield n.name
+        val expanded = for (k <- inner) yield col(colName + '.' + k).alias(colName + '_' + k)
+        df.select(col("*") +: expanded: _*).drop(colName)
+      }
+      case a: ArrayType => df.withColumn(colName, explode_outer(col(colName)))
+    }
+    flattenDf(newDf)
+  }
```

Comment:
And some Scala/Spark remarks here
- this gets all complex fields, but only process one field per call of flattenDf. Performancewise it would be better to handle all fields of a level at the same time.
- it creates many select statements, which is not optimal for Sparks logical plan. It would be better to add new columns using withColumn together with foldLeft(df){...}

Here a suggestion:
```suggestion
  private def flattenDf(df: DataFrame): DataFrame = {
    val complexFields = getComplexFields(df)
    complexFields.foldLeft(df){
      case (df, (colName, colType)) =>
        val newDf = colType match {
          case StructType(fields) =>
            fields.map(_.name).foldLeft(df) {
              case (df, name) => df.withColumn(colName + '_' + name, col(colName + '.' + name))
            }.drop(colName)
          case a: ArrayType => df.withColumn(colName, explode_outer(col(colName)))
        }
        flattenDf(newDf)
    }
  }
```

Reply:
Thanks @zzeekk I like the idea, but this would remove the tail-recursive structure and calls the function flattenDf for each column. Do you think the tradeoff is still worth it?